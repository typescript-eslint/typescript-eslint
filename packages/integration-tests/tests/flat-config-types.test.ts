import {
  eslintIntegrationTest,
  typescriptIntegrationTest,
} from '../tools/integration-test-base';

typescriptIntegrationTest(
  __filename,
  ['--allowJs', '--esModuleInterop', 'eslint.config.js'],
  out => {
    const lines = out
      .split('\n')
      .filter(
        line =>
          // error TS18028: Private identifiers are only available when targeting ECMAScript 2015 and higher.
          // this is fine for us to ignore in this context
          !line.includes('error TS18028'),
      )
      .join('\n');

    // The stylistic type errors: https://github.com/eslint-stylistic/eslint-stylistic/issues/276
    expect(lines).toMatchInlineSnapshot(`
      "node_modules/@stylistic/eslint-plugin-plus/dts/index.d.ts(7,46): error TS2694: Namespace '"/<tmp_folder>/node_modules/@types/eslint/index".ESLint' has no exported member 'RuleModule'.
      node_modules/@stylistic/eslint-plugin/dist/dts/rule-options.d.ts(6,11): error TS2320: Interface 'UnprefixedRuleOptions' cannot simultaneously extend types 'UnprefixedRuleOptions' and 'UnprefixedRuleOptions'.
        Named property ''comma-dangle'' of types 'UnprefixedRuleOptions' and 'UnprefixedRuleOptions' are not identical.
      node_modules/@stylistic/eslint-plugin/dist/dts/rule-options.d.ts(6,11): error TS2320: Interface 'UnprefixedRuleOptions' cannot simultaneously extend types 'UnprefixedRuleOptions' and 'UnprefixedRuleOptions'.
        Named property ''keyword-spacing'' of types 'UnprefixedRuleOptions' and 'UnprefixedRuleOptions' are not identical.
      node_modules/@stylistic/eslint-plugin/dist/dts/rule-options.d.ts(6,11): error TS2320: Interface 'UnprefixedRuleOptions' cannot simultaneously extend types 'UnprefixedRuleOptions' and 'UnprefixedRuleOptions'.
        Named property ''lines-around-comment'' of types 'UnprefixedRuleOptions' and 'UnprefixedRuleOptions' are not identical.
      node_modules/@stylistic/eslint-plugin/dist/dts/rule-options.d.ts(6,11): error TS2320: Interface 'UnprefixedRuleOptions' cannot simultaneously extend types 'UnprefixedRuleOptions' and 'UnprefixedRuleOptions'.
        Named property ''lines-between-class-members'' of types 'UnprefixedRuleOptions' and 'UnprefixedRuleOptions' are not identical.
      node_modules/@stylistic/eslint-plugin/dist/dts/rule-options.d.ts(6,11): error TS2320: Interface 'UnprefixedRuleOptions' cannot simultaneously extend types 'UnprefixedRuleOptions' and 'UnprefixedRuleOptions'.
        Named property ''padding-line-between-statements'' of types 'UnprefixedRuleOptions' and 'UnprefixedRuleOptions' are not identical.
      "
    `);
  },
);
eslintIntegrationTest(__filename, 'eslint.config.js', true);
