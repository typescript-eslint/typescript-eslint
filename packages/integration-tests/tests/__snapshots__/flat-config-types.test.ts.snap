// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flat-config-types eslint should work successfully 1`] = `
[
  {
    "errorCount": 3,
    "fatalErrorCount": 0,
    "filePath": "<root>/eslint.config.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": [
      {
        "column": 10,
        "endColumn": 21,
        "endLine": 38,
        "line": 38,
        "message": "'_otherCases' is defined but never used.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "@typescript-eslint/no-unused-vars",
        "severity": 2,
      },
      {
        "column": 5,
        "endColumn": 24,
        "endLine": 55,
        "line": 55,
        "message": "Include a description after the "@ts-expect-error" directive to explain why the @ts-expect-error is necessary. The description must be 3 characters or longer.",
        "messageId": "tsDirectiveCommentRequiresDescription",
        "nodeType": "Line",
        "ruleId": "@typescript-eslint/ban-ts-comment",
        "severity": 2,
      },
      {
        "column": 5,
        "endColumn": 24,
        "endLine": 60,
        "line": 60,
        "message": "Include a description after the "@ts-expect-error" directive to explain why the @ts-expect-error is necessary. The description must be 3 characters or longer.",
        "messageId": "tsDirectiveCommentRequiresDescription",
        "nodeType": "Line",
        "ruleId": "@typescript-eslint/ban-ts-comment",
        "severity": 2,
      },
    ],
    "output": "// @ts-check

import { FlatCompat } from '@eslint/eslintrc'
import eslint from '@eslint/js'
import stylisticPlugin from '@stylistic/eslint-plugin'
import deprecationPlugin from 'eslint-plugin-deprecation'
import jestPlugin from 'eslint-plugin-jest'
import tseslint from 'typescript-eslint'

import __dirname from './dirname.cjs'

const compat = new FlatCompat({
  baseDirectory: __dirname,
  recommendedConfig: {},
  allConfig: {},
})

// this config is run through eslint as part of the integration test
// so it needs to be a correct config
export default tseslint.config(
  {
    // config with just ignores is the replacement for \`.eslintignore\`
    ignores: ['**/build/**', '**/dist/**', 'src/some/file/to/ignore.ts'],
  },
  {
    plugins: {
      ['@typescript-eslint']: tseslint.plugin,
      ['deprecation']: deprecationPlugin,
      ['jest']: jestPlugin,
    },
  },
  eslint.configs.recommended,
  ...tseslint.configs.recommended,
  stylisticPlugin.configs['recommended-flat'],
)

// wrapped in a function so they aren't executed at lint time
function _otherCases() {
  // these are just tests for the types and are not seen by eslint so they can be whatever
  tseslint.config({
    plugins: {
      ['@stylistic']: stylisticPlugin,
      ['@typescript-eslint']: tseslint.plugin,
      ['deprecation']: deprecationPlugin,
      ['jest']: jestPlugin,
    },
  })
  tseslint.config(
    eslint.configs.recommended,
    ...tseslint.configs.recommended,
    stylisticPlugin.configs['recommended-flat'],
    jestPlugin.configs['flat/recommended'],
  )
  tseslint.config(
    // @ts-expect-error
    compat.config(deprecationPlugin.configs.recommended),
    ...compat.config(jestPlugin.configs.recommended),
  )
  tseslint.config(
    // @ts-expect-error
    deprecationPlugin.configs.recommended,
    // this should error but doesn't because there are no types exported from the jest plugin
    jestPlugin.configs.recommended,
    // this should error but doesn't because there are no types exported from the jest plugin
    ...jestPlugin.configs['flat/recommended'],
  )
}
",
    "suppressedMessages": [],
    "usedDeprecatedRules": [
      {
        "replacedBy": [],
        "ruleId": "no-extra-semi",
      },
      {
        "replacedBy": [],
        "ruleId": "no-mixed-spaces-and-tabs",
      },
    ],
    "warningCount": 0,
  },
]
`;
